x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java

```diff
@@ -136,14 +148,19 @@ private static Decision debugYesAllowed(RoutingAllocation allocation, List<Strin
      * in order to know if there are planned topology changes in the cluster
      * that can remove a tier that's part of the cluster now.
      */
-    public static Optional<String> preferredAvailableTier(List<String> prioritizedTiers, DiscoveryNodes nodes, DesiredNodes desiredNodes) {
+    public static Optional<String> preferredAvailableTier(
+        List<String> prioritizedTiers,
+        DiscoveryNodes nodes,
+        DesiredNodes desiredNodes,
+        NodesShutdownMetadata shutdownMetadata
+    ) {
         final var desiredNodesPreferredTier = getPreferredTierFromDesiredNodes(prioritizedTiers, nodes, desiredNodes);
 
         if (desiredNodesPreferredTier.isPresent()) {
             return desiredNodesPreferredTier;
         }
 
-        return getPreferredAvailableTierFromClusterMembers(prioritizedTiers, nodes);
+        return getPreferredAvailableTierFromClusterMembers(prioritizedTiers, filterRemovingNodes(nodes, shutdownMetadata));
     }
```

Comment:
I am surprised `DataTierAllocationDecider#allocationAllowed(java.lang.String, org.elasticsearch.cluster.node.DiscoveryNode)` is not checking for the node shutdowns, is it already covered by some other decider?

Reply:
It should be covered by `NodeShutdownAllocationDecider` as you pointed out.

Reply:
Agreed, that case is covered by `NodeShutdownAllocationDecider`. We tried to keep the logic as isolated as possible, but in this case this decider needs to know about node shutdown so that it won't refuse to allocate shards with a tier preference to nodes without that tier when all the nodes in that tier are shutting down.